java/maven.hints/src/org/netbeans/modules/maven/hints/pom/UseReleaseOptionHint.java

```diff
@@ -0,0 +1,241 @@
+    "DESC_UseReleaseVersionHint=Matching source/target options can be replaced with a single release option (JEP 247). This enforces API compatibility with the chosen Java release.",
+    "FIX_UseReleaseVersionHint=Convert to release option for strict compatibility checks."})
+public class UseReleaseOptionHint implements POMErrorFixProvider {
+
+    private static final String TARGET_TAG = "target";
+    private static final String SOURCE_TAG = "source";
+    private static final String RELEASE_TAG = "release";
+
+    private static final Configuration config = new Configuration(UseReleaseOptionHint.class.getName(),
+                TIT_UseReleaseVersionHint(), DESC_UseReleaseVersionHint(), true, Configuration.HintSeverity.WARNING);
+
+    @Override
+    public List<ErrorDescription> getErrorsForDocument(POMModel model, Project prj) {
+
+        List<ErrorDescription> hints = new ArrayList<>();
+
+        Build build = model.getProject().getBuild();
+        if (build != null) {
+            for (Plugin plugin : build.getPlugins()) {
+                if (plugin.getArtifactId().equals("maven-compiler-plugin")) {
+                    if (!isPluginCompatible(plugin)) {
+                        return Collections.emptyList();
+                    }
+                    if (plugin.getConfiguration() != null) {
+                        hints.addAll(createHintsForParent("", plugin.getConfiguration()));
+                    }
+                    if (plugin.getExecutions() != null) {
+                        for (PluginExecution execution : plugin.getExecutions()) {
+                            if (execution.getConfiguration() != null) {
+                                hints.addAll(createHintsForParent("", execution.getConfiguration()));
+                            }
+                        }
+                    }
+                    break;
+                }
+            }
+        }
+
+        Properties properties = model.getProject().getProperties();
+        if (properties != null) {
+            hints.addAll(createHintsForParent("maven.compiler.", properties));
+        }
+
+        return hints;
+    }
+
+    private List<ErrorDescription> createHintsForParent(String prefix, POMComponent parent) {
+
+        int source;
+        int target;
+
+        // property name or an int value
+        String release = null;
+
+        try {
+            String sourceText = parent.getChildElementText(POMQName.createQName(prefix+SOURCE_TAG, true));
+            if (isPropertyName(sourceText)) {
+                release = sourceText;
+                sourceText = getProperty(sourceText, parent.getModel());
+            }
+
+            String targetText = parent.getChildElementText(POMQName.createQName(prefix+TARGET_TAG, true));
+            if (isPropertyName(targetText)) {
+                release = targetText;
+                targetText = getProperty(targetText, parent.getModel());
```

Comment:
Is this a good idea? If I read this correctly, this evaluates a property. My assumption would be, that there is a reason the user does not use the default properties from maven compiler plugin. If so you can't assume that the property will always hold the same value.

Reply:
it evaluates the property to check what is in it (9+ check a few lines below). But it remembers the property String. So when it adds the release tag it will use the property again, not the int value.